File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/submodel/metamodel/map/submodelelement/event/BasicEvent.java
package org.eclipse.basyx.submodel.metamodel.map.submodelelement.event;

import java.util.Map;

import org.eclipse.basyx.submodel.metamodel.api.reference.IReference;
import org.eclipse.basyx.submodel.metamodel.api.reference.enums.KeyElements;
import org.eclipse.basyx.submodel.metamodel.api.submodelelement.event.IBasicEvent;
import org.eclipse.basyx.submodel.metamodel.map.modeltype.ModelType;
import org.eclipse.basyx.submodel.metamodel.map.reference.Reference;
import org.eclipse.basyx.submodel.metamodel.map.submodelelement.SubmodelElement;

/**
 * A BasicEvent element as defined in DAAS document
 * 
 * @author conradi
 *
 */
public class BasicEvent extends SubmodelElement implements IBasicEvent {

	public static final String MODELTYPE = "BasicEvent";
	public static final String OBSERVED = "observed";
	
	public BasicEvent() {
		// Add model type
		putAll(new ModelType(MODELTYPE));
	}
	
	public BasicEvent(IReference observed) {
		this();
		put(OBSERVED, observed);
	}
	
	@Override
	protected KeyElements getKeyElement() {
		return KeyElements.BASICEVENT;
	}
	
	/**
	 * Creates a BasicEvent object from a map
	 * 
	 * @param obj a BasicEvent object as raw map
	 * @return a BasicEvent object, that behaves like a facade for the given map
	 */
	public static BasicEvent createAsFacade(Map<String, Object> obj) {
		BasicEvent facade = new BasicEvent();
		facade.setMap(obj);
		return facade;
	}
	
	/**
	 * Returns true if the given submodel element map is recognized as an BasicEvent element
	 */
	public static boolean isBasicEvent(Map<String, Object> map) {
		String modelType = ModelType.createAsFacade(map).getName();
		// Either model type is set or the element type specific attributes are contained (fallback)
		return MODELTYPE.equals(modelType) || (modelType == null && map.containsKey(OBSERVED));
	}

	@Override
	@SuppressWarnings("unchecked")
	public IReference getObserved() {
		return Reference.createAsFacade((Map<String, Object>) get(OBSERVED));
	}
}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/map_v2/submodelelement/BasicEvent.cpp
#include <BaSyx/submodel/map_v2/submodelelement/BasicEvent.h>

namespace basyx {
namespace submodel {
namespace map {

constexpr char BasicEvent::Path::Observed[];

BasicEvent::BasicEvent(const std::string & idShort, Reference observed, ModelingKind kind)
  : SubmodelElement{idShort, kind}
  , observed{observed}
{
  this->map.insertKey(Path::Observed, this->observed.getMap());
}

BasicEvent::BasicEvent(basyx::object obj)
  : SubmodelElement{obj}
  , observed{obj.getProperty(Path::Observed)}
{
  this->map.insertKey(Path::Observed, this->observed.getMap());
}

const api::IReference & BasicEvent::getObserved() const
{
  return this->observed;
}

}
}
}
